pkg/envoy/envoy.go

```diff
@@ -150,18 +151,35 @@ func StartEnvoy(adminPort uint32, stateDir, logPath string, baseID uint64) *Envo
 	// case no one reader reads it.
 	started := make(chan bool, 1)
 	go func() {
-		logger := &lumberjack.Logger{
-			Filename:   logPath,
-			MaxSize:    100, // megabytes
-			MaxBackups: 3,
-			MaxAge:     28,   //days
-			Compress:   true, // disabled by default
+		var logger *lumberjack.Logger
+		var logFormat string
+		if logPath != "" {
+			// Use the Envoy default log format when logging to a separate file
+			logFormat = "[%Y-%m-%d %T.%e][%t][%l][%n] %v"
+			logger = &lumberjack.Logger{
+				Filename:   logPath,
+				MaxSize:    100, // megabytes
+				MaxBackups: 3,
+				MaxAge:     28,   //days
+				Compress:   true, // disabled by default
+			}
+			defer logger.Close()
+		} else {
+			// Use log format that looks like Cilium logs when integrating logs
+			// The logs will be reported as coming from the cilium-agent, so
+			// we add the thread id to be able to differentiate between Envoy's
+			// main and worker threads.
+			logFormat = "level=%l msg=\"%v\" subsys=envoy-%n thread=%t"
 		}
-		defer logger.Close()
 		for {
-			cmd := exec.Command("cilium-envoy", "-l", mapLogLevel(log.Level), "-c", bootstrapPath, "--base-id", strconv.FormatUint(baseID, 10))
-			cmd.Stderr = logger
-			cmd.Stdout = logger
+			cmd := exec.Command("cilium-envoy", "-l", mapLogLevel(log.Level), "-c", bootstrapPath, "--base-id", strconv.FormatUint(baseID, 10), "--log-format", logFormat)
+			if logger != nil {
+				cmd.Stderr = logger
+				cmd.Stdout = logger
+			} else {
+				cmd.Stderr = os.Stderr
```

Comment:
This is not correct. We can't assume we're logging into stdout / stderr. We have support for Fluentd, etc. in Cilium and we can't break that. And we want to make sure that we blend well with the other logs from Cilium. So those logs must go through logrus.

The right way to do this is to use logrus's `Writer` support:
```
logger := log.WithFields(...).WriterLevel(logrus.InfoLevel)
...
cmd.Stderr = logger
cmd.Stdout = logger
```

That's not great, because we'll want to set the logrus log level according to the Envoy log level. Here even debug logs will be logged at Info level.
But that's a good start.

It may be nice to have different log levels for stdout and stderr:
```
stderrLogger := log.WithFields(...).WriterLevel(logrus.WarnLevel)
stdoutLogger := log.WithFields(...).WriterLevel(logrus.InfoLevel)
...
cmd.Stderr = stderrLogger
cmd.Stdout = stdoutLogger
```

Reply:
Thanks for the pointers, will start with this and see if I can get the the log levels from Envoy to logrus.